src/Microsoft.Azure.WebJobs.Host/Bindings/OpenType.cs

```diff
@@ -6,28 +6,328 @@
+                        }
+
+                        name += TypeToString(arg);
+                        i++;
+                    }
+                    name += ">";
+                    return name;
+                }
+                return t.Name;
+            }
+        }
+
+        // Match a generic type with 1 generic arg. 
+        // like IEnumerable<T>,  IQueryable<T>, etc. 
+        private class SingleGenericArgOpenType : OpenType
+        {
+            private readonly OpenType _inner;
+            private readonly Type _outerType;
+
+            public SingleGenericArgOpenType(Type outerType, OpenType inner)
+            {
+                _inner = inner;
+                _outerType = outerType;
+            }
+
+            public override bool IsMatch(Type type, OpenTypeMatchContext context)
+            {
+                if (type == null)
+                {
+                    throw new ArgumentNullException(nameof(type));
+                }
+                if (context == null)
+                {
+                    throw new ArgumentNullException(nameof(context));
+                }
+                if (type.IsGenericType &&
+                    type.GetGenericTypeDefinition() == _outerType)
+                {
+                    var args = type.GetGenericArguments();
+
+                    return _inner.IsMatch(args[0], context);
+                }
+
+                return false;
+            }
+         
+            internal override string GetDisplayName()
+            {
+                var name = _outerType.GetGenericTypeDefinition().Name;
+                return name + "<" + _inner.GetDisplayName() + ">";
+            }
+        }
+
+        // Matches any T[] 
+        private class ArrayOpenType : OpenType
+        {
+            private readonly OpenType _inner;
+            public ArrayOpenType(OpenType inner)
+            {
+                _inner = inner;
+            }
+
+            public override bool IsMatch(Type type, OpenTypeMatchContext context)
+            {
+                if (type == null)
+                {
+                    throw new ArgumentNullException(nameof(type));
+                }
+                if (context == null)
+                {
+                    throw new ArgumentNullException(nameof(context));
+                }
+                
+                if (type.IsArray)
+                {
+                    var elementType = type.GetElementType();
+                    return _inner.IsMatch(elementType, context);
+                }
+                return false;
+            }
+
+            internal override string GetDisplayName()
+            {
+                return _inner.GetDisplayName() + "[]";
+            }
+        }
+                 
+        // Formalize definition of a "poco"
+        // This can be used for general "T --> JObject" bindings. 
+        // The exact definition here comes from the WebJobs v1.0 Queue binding.
+        public class Poco : OpenType
```

Comment:
PocoOpenType, or perhaps CustomOpenType?

Reply:
This is a nested type right now (more discoverable), so it's "OpenType.Poco".   Happy to rename (or make unnested). 


---
In reply to: [150931977](https://github.com/Azure/azure-webjobs-sdk/pull/1422#discussion_r150931977) [](ancestors = 150931977)